File: results/smoke-claude-code-task_02_admin_tool-terse-NI-run5/output_files/src/auth/oidc.ts
/**
 * OpenID Connect authentication strategy (Authorization Code + PKCE).
 *
 * Activated only when OIDC_ENABLED=true. The issuer is discovered at startup
 * via the IdP's /.well-known/openid-configuration document.
 */
import { Issuer, Strategy as OidcStrategy, type TokenSet, type UserinfoResponse } from 'openid-client';
import { config } from '../config';
import { logger } from '../logger';
import { provisionFromSso } from './provisioning';

export async function buildOidcStrategy(): Promise<OidcStrategy<Record<string, unknown>>> {
  if (!config.OIDC_ISSUER_URL || !config.OIDC_CLIENT_ID || !config.OIDC_CLIENT_SECRET) {
    throw new Error('OIDC_ENABLED is true but OIDC_ISSUER_URL / CLIENT_ID / CLIENT_SECRET are missing');
  }

  const issuer = await Issuer.discover(config.OIDC_ISSUER_URL);
  logger.info({ issuer: issuer.metadata.issuer }, 'OIDC issuer discovered');

  const client = new issuer.Client({
    client_id: config.OIDC_CLIENT_ID,
    client_secret: config.OIDC_CLIENT_SECRET,
    redirect_uris: [config.oidcCallbackUrl],
    response_types: ['code'],
  });

  const verify = async (
    tokenSet: TokenSet,
    userinfo: UserinfoResponse,
    done: (err: unknown, user?: Record<string, unknown>) => void,
  ) => {
    try {
      const claims = { ...tokenSet.claims(), ...userinfo };
      const email = String(claims.email ?? '');
      const user = await provisionFromSso({
        provider: 'OIDC',
        subject: String(claims.sub ?? email),
        email,
        name: String(claims.name ?? claims.preferred_username ?? email),
      });
      done(null, user as unknown as Record<string, unknown>);
    } catch (err) {
      logger.warn({ err }, 'OIDC login rejected');
      done(err);
    }
  };

  return new OidcStrategy(
    {
      client,
      params: { scope: config.OIDC_SCOPE },
      usePKCE: true,
    },
    verify,
  );
}

File: results/smoke-claude-code-task_02_admin_tool-terse-NI-run1/output_files/app/auth/oidc.py
from authlib.integrations.starlette_client import OAuth

from app.config import get_settings

settings = get_settings()
oauth = OAuth()

if settings.oidc_client_id and settings.oidc_discovery_url:
    oauth.register(
        name="oidc",
        client_id=settings.oidc_client_id,
        client_secret=settings.oidc_client_secret,
        server_metadata_url=settings.oidc_discovery_url,
        client_kwargs={"scope": "openid email profile groups"},
    )


def claims_to_user_info(claims: dict) -> dict:
    """Normalize OIDC claims into a common shape."""
    return {
        "subject": claims.get("sub", ""),
        "email": claims.get("email", "").lower(),
        "name": claims.get("name") or claims.get("preferred_username", ""),
        "groups": claims.get("groups") or claims.get("roles") or [],
        "provider": "oidc",
    }
